src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -119,11 +119,22 @@ internal JsonClassInfo(Type type, JsonSerializerOptions options)
                     // Add a single property that maps to the class type so we can have policies applied.
                     JsonPropertyInfo policyProperty = AddPolicyProperty(type, options);
 
-                    // Use the type from the property policy to get any late-bound concrete types (from an interface like IDictionary).
-                    CreateObject = options.ClassMaterializerStrategy.CreateConstructor(policyProperty.RuntimePropertyType);
+                    Type elementType = GetElementType(type);
+
+                    if (!DefaultImmutableConverter.TypeIsImmutableDictionary(type))
+                    {
+                        // Use the type from the property policy to get any late-bound concrete types (from an interface like IDictionary).
+                        CreateObject = options.ClassMaterializerStrategy.CreateConstructor(policyProperty.RuntimePropertyType);
+                    }
+                    else
+                    {
+                        // Since immutable dictionaries do not implement Dictionary<,>, the runtime property type could not be set to
+                        // Dictionary<string, TElement>, we have to set CreateObject this way.
+                        CreateObject = options.ClassMaterializerStrategy.CreateConstructor(
+                            typeof(Dictionary<,>).MakeGenericType(typeof(string), elementType));
```

Comment:
This is a little surprising to me. Is this ever called? I haven't dug into the serializer code enough to really know what this is for  (referring to the comment at L119-120).

Reply:
Yes, it's called here: https://github.com/dotnet/corefx/blob/39216c8d8250f9af8acd6ce8a8fe1b7ac6754317/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs#L75

`policyPropertyInfo` describes how to (de)serialize the type. It contains the declared property type and the runtime property type. For the scope of this discussion, we set the runtime property type and use it to create the object to be populated (and either returned directly, or passed as input to a converter that constructs and returns the desired object).

For non-immutable dictionaries, we set the runtime type (to be returned) to a concrete `Dictionary<string, TElement>` [here](https://github.com/dotnet/corefx/blob/39216c8d8250f9af8acd6ce8a8fe1b7ac6754317/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs#L34).

[The runtime property type needs to implement the declared property type](https://github.com/dotnet/corefx/blob/39216c8d8250f9af8acd6ce8a8fe1b7ac6754317/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs#L16), but in this case, immutable dictionaries do not implement `Dictionary<,>`, so we leave the runtime property type as is `((I)ImmutableXDictionary<string, TElement>)`.

For immutable dicts, we
- forgo the runtime property type
- manually create a concrete dictionary type
- create and cache a delegate for the type's constructor, and
- later on create an object with the delegate and use it as a temp to be passed as an argument to the appropriate immutable converter